### PR TITLE
server: when wan federating via mesh gateways only do heuristic primary DC bypass on the leader

### LINKED ISSUE #9341: wanfed via mesh gateways on follower nodes in a secondary DC lack a fed state replicator feedback loop
On a non-leader node in a secondary DC nothing is going to be available to invoke this method:
https://github.com/hashicorp/consul/pull/7931/files#diff-cef5078cde0c5594853fc804b484520e264d323995c4d75f449ba65b2d47d669R76

This will lead to all follower nodes in a secondary DC bypassing their local mesh gateways when talking to the leader.

The fix should be to just have followers *only* use the local gateways, and only have the leader use the federation state replicator loop to bypass configuration hiccups and rebootstrap.

Affects 1.8+

### PR DESCRIPTION
**Pull request by rboyer**
Fixes #9341

### CODE REVIEW

File: agent/consul/gateway_locator.go [hunk 1]
```diff
@@ -140,6 +151,12 @@ func (g *GatewayLocator) DialPrimaryThroughLocalGateway() bool {
 const localFederationStateReplicatorFailuresBeforeDialingDirectly = 3
 
 func (g *GatewayLocator) dialPrimaryThroughLocalGateway() bool {
+	if !g.useReplicationSignal {
+		// Followers should blindly assume these gateways work. The leader will
+		// try to bypass them and correct the replicated fedreation state info
```
Comment (rboyer) [inline agent/consul/gateway_locator.go:156]:
```suggestion
		// try to bypass them and correct the replicated federation state info
```

File: agent/consul/gateway_locator.go [hunk 2]
```diff
@@ -74,9 +75,19 @@ type GatewayLocator struct {
...
+	if g == nil {
+		return
+	}
+
 	g.lastReplLock.Lock()
 	defer g.lastReplLock.Unlock()
+
+	if fromReplication {
+		// If we get info from replication, assume replication is operating.
+		g.useReplicationSignal = true
```
Comment (freddygv) [inline agent/consul/gateway_locator.go:88]: What do we mean by replication is operating? Just that there's a Replicator routine running?

Where I'm lost is that there could have been an error in this call to `fetchRemote`:
https://github.com/hashicorp/consul/pull/9366/files#diff-1b7bee896bb2362c8062a2d6ee0f449799817d428f000c8a5c6347e5a041f59aR42

Comment (rboyer) [reply agent/consul/gateway_locator.go:88]: Sigh. While on the circular quest to fix this series of wanfed adverse interactions I completely missed that the `err` coming into this function can be `nil` when plumbing this up.

Comment (rboyer) [reply agent/consul/gateway_locator.go]: It's supposed to be "successfully did a remote RPC"

Comment (rboyer) [reply agent/consul/gateway_locator.go]: Fixed.

**Review by freddygv — APPROVED** [review]
:shipit:

**Pull request merged**